Allow ports to be reused in gloo

H-Huang opened the pull request:
Summary:
ProcessGroupGloo and gloo seem to be opening and closing sockets without allowing the port to be reused. We see this issue pop up in larger training jobs "Address already in use" and we assume it to be because all the ephemeral ports are exhausted.

This diff allows ports to be reused, we see a reduced number of ports being in `TIME_WAIT` state.

context: https://fb.workplace.com/groups/319878845696681/permalink/5988899781205532/

another issue: https://fb.workplace.com/groups/319878845696681/permalink/958768178474408/

Test Plan: Add a gloo test to create 4 groups of size 64 using multithreaded PG + gloo. In total 256 ranks.

Differential Revision: D44029927

kwen2501 commented:
Thanks for looking into this issue @H-Huang !

> we assume it to be because all the ephemeral ports are exhausted.

Curious -- do we know what caused this exhaustion?

On the change, IIRC, `SO_REUSEADDR` might have some risk. 
This blog has some discussion:
https://stackoverflow.com/questions/3229860/what-is-the-meaning-of-so-reuseaddr-setsockopt-option-linux

Do we know if it would be safe for it to be the default here? I haven't dug deep into how the call stack here is triggered.

H-Huang committed:
Allow ports to be reused in gloo (#97677)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/97677

X-link: https://github.com/facebookincubator/gloo/pull/353

ProcessGroupGloo and gloo seem to be opening and closing sockets without allowing the port to be reused. We see this issue pop up in larger training jobs "Address already in use" and we assume it to be because all the ephemeral ports are exhausted.

This diff allows ports to be reused, we see a reduced number of ports being in `TIME_WAIT` state.

context: https://fb.workplace.com/groups/319878845696681/permalink/5988899781205532/

another issue: https://fb.workplace.com/groups/319878845696681/permalink/958768178474408/

Test Plan: Add a gloo test to create 4 groups of size 64 using multithreaded PG + gloo. In total 256 ranks.

Differential Revision: D44029927

fbshipit-source-id: 9c31c38485333602c33e12c12813bea33ccb9438
